content/blog/explicit-apis-vs-magic-directives.mdx

```diff
@@ -0,0 +1,408 @@
+
+Directives can't give you this. They're processed at build time—after TypeScript has already checked your code.
+
+### 2. Testing Works Without Framework Knowledge
+
+If testing requires understanding framework transformations, teams won't write tests. With directives, you're testing transformed code. With libraries, you're testing what you wrote.
+
+Here's testing with explicit APIs:
+
+```tsx
+const mockStep = {
+  run: jest.fn((id, fn) => fn()),
+  sleep: jest.fn(),
+  waitForEvent: jest.fn()
+};
+
+await processOrder({ event: mockEvent, step: mockStep });
+expect(mockStep.run).toHaveBeenCalledWith("validate", expect.any(Function));
+```
+
+You're mocking an object. Any developer can read this. No special framework knowledge required. No need to understand what the compiler did to your code.
+
+### 3. Debugging Shows Your Actual Code
+
+When production breaks, you need clear visibility into what happened.
+
+**With directives:** Stack traces point to transformed code. Breakpoints might not align with your source. You're debugging something you didn't write. Good luck explaining that to your team at 3 AM.
+
+**With explicit libraries:** Stack traces show your actual code. Breakpoints work exactly where you set them. The debugger displays what you wrote. You can trace execution through the exact functions you deployed.
+
+This matters when you're diagnosing why a payment failed or why a notification didn't send. You need to see *your* code, not what the compiler generated.
+
+### 4. Safe Deployments Require Stable State Keys
+
+This was a painful lesson, but one of the most important. Here's what happens with implicit IDs when someone refactors while a sleep is happening:
+
+Code Version 1 runs as expected
+
+```tsx
+export async function handleUserSignup(email: string) {
+  "use workflow";
+  const user = await createUser(email);
+  await sendWelcomeEmail(user);
+  await sleep("10s");
+}
+```
+
+With the following result:
+
+<img src="/assets/blog/explicit-apis/first.png" alt="Steps with directives breaks" />
+
+Now, let’s look at what happens when a developer makes the following code change *while the sleep is happening:*
+
+```tsx
+export async function handleUserSignup(email: string) {
+  "use workflow";
+  await createOrg();
+  const user = await createUser(email);
+  await sendWelcomeEmail(user);
+  await sleep("10s");
+}
+```
+
+The result, when you manually upgrade the run to use new code:
+
+<img src="/assets/blog/explicit-apis/dupe.png" alt="Steps with directives breaks" />
+
+Workflows that were sleeping when you deployed your change are now corrupted. Because Vercel’s WDK steps don't have an explicit ID, they have to rely on “magic” to understand which steps have already been executed. Even tiny, seemingly inconsequential code changes can break its state, leading to re-executed logic. 
```

Comment:
```suggestion
Workflows that were sleeping when you deployed your change are now corrupted. Because WDK steps don't have an explicit ID, they have to rely on “magic” to understand which steps have already been executed. Even tiny, seemingly inconsequential code changes can break its state, leading to re-executed logic. 
```